p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		scope.Done()
+		if pc != nil {
+			_ = pc.Close()
+		}
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (l *listener) setupConnection(ctx context.Context, scope network.ConnManagementScope, remoteMultiaddr ma.Multiaddr, addr candidateAddr) (*webrtc.PeerConnection, tpt.CapableConn, error) {
+
+	settingEngine := webrtc.SettingEngine{}
+	settingEngine.SetAnsweringDTLSRole(webrtc.DTLSRoleServer)
+	settingEngine.SetICECredentials(addr.ufrag, addr.ufrag)
+	settingEngine.SetLite(true)
+	settingEngine.SetICEUDPMux(l.mux)
+	settingEngine.DisableCertificateFingerprintVerification(true)
+	settingEngine.DetachDataChannels()
+
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(settingEngine))
+
+	pc, err := api.NewPeerConnection(l.config)
+	if err != nil {
+		return pc, nil, err
+	}
+
+	signalChan := make(chan error)
+	var wrappedChannel *dataChannel
+	var handshakeOnce sync.Once
+
+	handshakeChannel, err := pc.CreateDataChannel("", &webrtc.DataChannelInit{
+		Negotiated: func(v bool) *bool { return &v }(true),
+		ID:         func(v uint16) *uint16 { return &v }(0),
+	})
+	if err != nil {
+		return pc, nil, err
+	}
+
+	handshakeChannel.OnOpen(func() {
+		rwc, err := handshakeChannel.Detach()
+		if err != nil {
+			signalChan <- errDatachannel("could not detach", err)
+			return
+		}
+		wrappedChannel = newDataChannel(
+			nil,
+			handshakeChannel,
+			rwc,
+			pc,
+			l.localAddr,
+			addr.raddr,
+		)
+		handshakeOnce.Do(func() {
+			signalChan <- nil
+		})
+	})
+
+	// Checking the peerconnection state is not necessary in this case as any
+	// error caused while accepting will trigger the onerror callback of the
+	// handshake channel.
+	handshakeChannel.OnError(func(e error) {
+		handshakeOnce.Do(func() {
+			signalChan <- e
+
+		})
+	})
+
+	clientSdpString := renderClientSdp(sdpArgs{
+		Addr:  addr.raddr,
+		Ufrag: addr.ufrag,
+	})
+	clientSdp := webrtc.SessionDescription{SDP: clientSdpString, Type: webrtc.SDPTypeOffer}
+	pc.SetRemoteDescription(clientSdp)
+
+	answer, err := pc.CreateAnswer(nil)
+	if err != nil {
+		return pc, nil, err
+	}
+
+	err = pc.SetLocalDescription(answer)
+	if err != nil {
+		return pc, nil, err
+	}
+
+	// await datachannel moving to open state
+	select {
+	case <-ctx.Done():
+		return pc, nil, ctx.Err()
+	case err := <-signalChan:
+		if err != nil {
+			return pc, nil, errDatachannel("datachannel error", err)
+		}
+	}
+
+	// The connection is instantiated before performing the Noise handshake. This is
+	// to handle the case where the remote is faster and attempts to initiate a stream
+	// before the ondatachannel callback can be set.
+	conn := newConnection(
+		pc,
+		l.transport,
+		scope,
+		l.transport.localPeerId,
+		l.transport.privKey,
+		l.localMultiaddr,
+		"",
+		nil,
+		remoteMultiaddr,
+	)
+
+	// we do not yet know A's peer ID so accept any inbound
+	secureConn, err := l.transport.noiseHandshake(ctx, pc, wrappedChannel, "", crypto.SHA256, true)
```

Comment:
I don't think it's correct to hardcode SHA256 here.

Reply:
It's the minimum requirement in the spec https://github.com/libp2p/specs/tree/master/webrtc#addressing

Reply:
That's the minimum requirement, but that doesn't mean we should hardcode it. The fingerprint could have used a different hash algorithm, right? https://github.com/libp2p/specs/tree/master/webrtc#connection-security